Feature: Wireguard

chifac08 opened the pull request:
## Proposed Commit Message
```
Wireguard modules provides a dynamic interface for configuring Wireguard (as a peer or server) in an easy way. 
```

## Additional Context
This module takes care of..

- writing interface configuration files
- enabling and starting interfaces
- installing wireguard-tools package
- loading wireguard kernel module
- executing readiness probes

The idea behind readiness probes is to ensure Wireguard connectivity before continuing the cloud-init process. This could be useful if you need access to specific services like an internal APT Repository Server (e.g  Landscape) to install/update packages. 

## Test Steps
We implemented unit tests on best effort. For real testing we used the following user-data and meta-data file:

user-data:
```
wireguard:
  interfaces:
    - name: wg0
      config_path: "{{ ds.meta_data.wg0_config_path }}"
      content: |
        [Interface]
        PrivateKey = {{ ds.meta_data.wg_private_key }}
        Address = {{ ds.meta_data.wg0_address }}
        Table = off
        [Peer]
        PublicKey = {{ ds.meta_data.wg0_public_key }}
        Endpoint = {{ ds.meta_data.wg0_endpoint }}:{{ ds.meta_data.wg0_port }}
        AllowedIPs = {{ ds.meta_data.wg0_allowedips }}
        PersistentKeepalive = 25
    - name: wg1
      config_path: "{{ ds.meta_data.wg1_config_path }}"
      content: |
        [Interface]
        PrivateKey = {{ ds.meta_data.wg_private_key }}
        Address = {{ ds.meta_data.wg1_address }}
        Table = off
        [Peer]
        PublicKey = {{ ds.meta_data.wg1_public_key }}
        Endpoint = {{ ds.meta_data.wg1_endpoint }}:{{ ds.meta_data.wg1_port }}
        AllowedIPs = {{ ds.meta_data.wg1_allowedips }}
        PersistentKeepalive = 25
  readinessprobe:
    - 'curl https://webhook.site/id'
```


meta-data:
```
wg_private_key: <private_key>
wg0_config_path: /etc/wireguard/wg0.conf
wg0_address: 192.168.1.1/24
wg0_public_key: <public_key>
wg0_endpoint: host1.example.net
wg0_port: 51820
wg0_allowedips: 0.0.0.0/0
wg1_config_path: /etc/wireguard/wg1.conf
wg1_address: 192.168.2.1/24
wg1_public_key: <public_key>
wg1_endpoint: host2.example.net
wg1_port: 51820
wg1_allowedips: 0.0.0.0/0
```


## Checklist:
 - [x ] My code follows the process laid out in [the documentation](https://cloudinit.readthedocs.io/en/latest/topics/contributing.html)
 - [ x] I have updated or added any unit tests accordingly
 - [ x] I have updated or added any documentation accordingly

holmanb commented on cloudinit/config/cc_wireguard.py:
```diff
@@ -0,0 +1,243 @@
+    try:
+        LOG.debug("Running: systemctl enable wg-quick@%s", {wg_int["name"]})
+        subp.subp(
+            f'systemctl enable wg-quick@{wg_int["name"]}',
```
Drive by comment: Rather than assuming systemd, we should use `cloud.distro.manage_service()`.

chifac08 commented:
@aciba90  @holmanb 
Implemented the changes/suggestions you requested.

Please review again.

thx

holmanb commented on cloudinit/config/cc_wireguard.py:
```diff
@@ -0,0 +1,242 @@
+        LOG.debug("Found Wireguard section in config")
+        wg_section = cfg["wireguard"]
+
+    if wg_section is None:
```
Why allow an empty config that just gets skipped? I don't understand that. Instead could we alter the schema to not disallow `null`, and just skip this module if the `wireguard` key doesn't exist in the config?

chifac08 replied:
Totally agree with you. I am going to implement it that way.

holmanb commented on tests/unittests/config/test_cc_wireguard.py:
```diff
@@ -0,0 +1,245 @@
+    @pytest.mark.parametrize(
+        "config, error_msg",
+        [
+            # Allow empty wireguard config
```
Why?

holmanb reviewed (changes requested):
What's a readiness probe? I haven't used Wireguard in a couple of years, but I don't recall readiness probes at all and the top hits in a search engine seem to point to kubernetes-related topics. Based on the code this is an arbitrary command that blocks until the network is ready? Could you please add some more context for this?

chifac08 commented:
@holmanb 
**What's a readiness probe?** 
The idea behind readiness probes is to ensure Wireguard connectivity before continuing the cloud-init process. This could be useful if you need access to specific services like an internal APT Repository Server (e.g Landscape) to install/update packages. 

**Example:**
An edge device can't access the internet but uses cloud-init modules that will install packages (e.g landscape, packages, ubuntu_advantage). Those modules will fail due to missing internet connection. Our "wireguard" module fixes that problem as it waits until all readinessprobes (which can be arbitrary commands - e.g. checking if a proxy server is reachable over Wireguard network) are finished before continuing the cloud-init "config" stage.

The word "readinessprobe" is indeed a Kubernetes Feature and we thought the term also best fit's our use case.

If you have any further questions, please let me know


Additionally, it looks like CI tests failed. -> Suddenly my unit tests executed with "tox" do not fail but I'll have a look at it.

holmanb commented:
> @holmanb **What's a readiness probe?** ...

Perfect, thanks for the context. Could you please include this description or similar in the user docs and/or as comments in the examples?

aciba90 reviewed (changes requested):
> Additionally, it looks like CI tests failed. -> Suddenly my unit tests executed with "tox" do not fail but I'll have a look at it.

This is failing because the CI runner does not have `wireguard` installed. Mocking the call will fix CI.

chifac08 commented:
@holmanb @aciba90 
I implemented the changes you requested and improved documentation. Implemented a new function for loading wireguard kernel module as it seemed like the more beautiful way by separating package installation and loading kernel modules. 

I hope that my examples and documentation is committed on the right place. If not, I would be delighted if you could help me :)

Please review again.

thx

holmanb commented:
@chifac08 - Thanks for making those changes!

Apologies for the delayed review.

I'm a little conflicted about how we can ensure this module is updated if/when the required commands change. If the wireguard-tools invocation required changes, for example, in a way that would cause this module to stop working, we would ideally have an integration test that fails. I'm not exactly sure how best to do that.
 
To test both the configuration commands as well as the VPN tunnel would require a minimum of 2 instances, correct? I don't think we have synchronization tooling for that builtin to our integration testing code.

At a minimum, I think we should add a test that actually configures one node (even if it doesn't actually test a connection). Thoughts?

chifac08 commented:
@holmanb thx for reviewing

I really like the idea of implementing integration tests and here are my thoughts. 

Testing configuration commands can easily be done by (as you already said) configuring a node as 
such tests cover all needed wg-quick commands. Furthermore, wg-quick will throw an error message if 
wireguard config is invalid (which is also covered in single node tests). In my opinion, testing the 
VPN tunnel goes beyond the scope of this module as the "content" parameter can be dynamically set and 
there are too many variables (network, firewalling, etc). which makes testing very difficult. The 
"content" parameter makes the user responsible for writing a valid wireguard configuration instead of 
having the developer wrapping a wg conf as parameters. 

For example, you can easily bring up 2 wireguard peers without errors being thrown but when you try to ping each
other's IP Adress it won't work due to wrong public keys. wg-quick as well as the "wg show" command won't 
throw an error. "wg show" may indicate that you have sent some data but didn't recieve it on both hosts 
without seeing an "real" error message or exception.

I am going to write single node integration tests and head back to you when finished.

Please let me know your thoughts and questions.

All current integration tests are executed on "Bionic" right?

thx

holmanb commented:
> In my opinion, testing the VPN tunnel goes beyond the scope of this module as the "content" parameter can be dynamically set and there are too many variables (network, firewalling, etc). which makes testing very difficult.

+1 agreed

> I am going to write single node integration tests and head back to you when finished.

Excellent, thanks!
 
> All current integration tests are executed on "Bionic" right?

Yes

chifac08 commented:
@holmanb

Implemented integration tests with following test cases:
 - installed wireguard-tools package
 - loaded wireguard kernel module
 - wireguard config is written to file
 - wireguard config is detected as ASCII file
 - integrity of wireguard config files (hashes)
 - wireguard systemd service per interface is active
 - wireguard interface is pingable (readiness probe)

integration tests where executed for
lxd:
  vm:
    - bionic [**success**]
    - focal [**success**]
    - jammy [**success**]
  container:
    - bionic [**success**]
    - focal [**success**]
    - jammy [**success**]

Furthermore, i implemented a logic where "wireguard" meta package gets installed when kernel version is lower 5.6. The wireguard module was merged into kernel version 5.6. So Wireguard module is capable of os releases like Bionic.

Unfortunately, integration tests failed but i don't know why. I used the following command to execute integration tests and it worked well:
`CLOUD_INIT_PLATFORM="lxd_container" CLOUD_INIT_OS_IMAGE="bionic" CLOUD_INIT_KEEP_INSTANCE=True CLOUD_INIT_CLOUD_INIT_SOURCE="$(ls cloud-init_all.deb)" tox -e unit-tests tests/unittests/config/test_cc_wireguard.py`

Please have a look at it and let me know what you think.

thx

aciba90 commented on tests/integration_tests/modules/test_wireguard.py:
```diff
@@ -0,0 +1,90 @@
+    def test_wireguard(
+        self, cmd, expected_out, class_client: IntegrationInstance
+    ):
+        out = class_client.execute(cmd)
```
I think what we want to check that the command did not fail and that `expected_out` is in the stdout right?
```suggestion
        result = class_client.execute(cmd)
        assert result.ok
        assert expected_out in result.stdout
```

aciba90 reviewed (changes requested):
Thanks, @chifac08, for adding integration tests.

> Furthermore, i implemented a logic where "wireguard" meta package gets installed when kernel version is lower 5.6. The wireguard module was merged into kernel version 5.6. So Wireguard module is capable of os releases like Bionic.
> 
> Unfortunately, integration tests failed but i don't know why. I used the following command to execute integration tests and it worked well: `CLOUD_INIT_PLATFORM="lxd_container" CLOUD_INIT_OS_IMAGE="bionic" CLOUD_INIT_KEEP_INSTANCE=True CLOUD_INIT_CLOUD_INIT_SOURCE="$(ls cloud-init_all.deb)" tox -e integration-tests tests/integration_tests/modules/test_wireguard.py`
> 
> Please have a look at it and let me know what you think.
> 
> thx

When I add the following test:

```diff
diff --git a/tests/integration_tests/modules/test_wireguard.py b/tests/integration_tests/modules/test_wireguard.py
index 68134d7f..7f2b2dc7 100644
--- a/tests/integration_tests/modules/test_wireguard.py
+++ b/tests/integration_tests/modules/test_wireguard.py
@@ -6,6 +6,7 @@ TODO Description
 import pytest
 
 from tests.integration_tests.instances import IntegrationInstance
+from tests.integration_tests.util import verify_clean_log
 
 ASCII_TEXT = "ASCII text"
 
@@ -44,6 +45,11 @@ wireguard:
 @pytest.mark.ci
 @pytest.mark.user_data(USER_DATA)
 class TestWireguard:
+
+    def test_clean_log(self, class_client: IntegrationInstance):
+        log_content = class_client.read_from_file("/var/log/cloud-init.log")
+        verify_clean_log(log_content)
+
     @pytest.mark.parametrize(
         "cmd,expected_out",
         (
```

it shows up an error did occur trying to load the `wireguard` kernel module:

```txt
...
2022-07-21 08:49:08,710 - modules.py[DEBUG]: Running module wireguard (<module 'cloudinit.config.cc_wireguard' from '/usr/lib/python3/dist-packages/cloudinit/config/cc_wireguard.py'>) with frequency once-per-instance
2022-07-21 08:49:08,710 - handlers.py[DEBUG]: start: modules-config/config-wireguard: running config-wireguard with frequency once-per-instance
2022-07-21 08:49:08,710 - util.py[DEBUG]: Writing to /var/lib/cloud/instances/cloudinit-0721-0848573mbwo3jm/sem/config_wireguard - wb: [644] 23 bytes
2022-07-21 08:49:08,710 - helpers.py[DEBUG]: Running config-wireguard using lock (<FileLock using file '/var/lib/cloud/instances/cloudinit-0721-0848573mbwo3jm/sem/config_wireguard'>)
2022-07-21 08:49:08,710 - cc_wireguard.py[DEBUG]: Found Wireguard section in config
2022-07-21 08:49:08,710 - util.py[DEBUG]: Writing to /var/lib/cloud/instances/cloudinit-0721-0848573mbwo3jm/sem/update_sources - wb: [644] 24 bytes
2022-07-21 08:49:08,711 - helpers.py[DEBUG]: Running update-sources using lock (<FileLock using file '/var/lib/cloud/instances/cloudinit-0721-0848573mbwo3jm/sem/update_sources'>)
2022-07-21 08:49:08,711 - debian.py[DEBUG]: Waiting for apt lock
2022-07-21 08:49:08,711 - debian.py[DEBUG]: apt lock available
2022-07-21 08:49:08,711 - subp.py[DEBUG]: Running command ['eatmydata', 'apt-get', '--option=Dpkg::Options::=--force-confold', '--option=Dpkg::options::=--force-unsafe-io', '--assume-yes', '--quiet', 'update'] with allowed return codes [0] (shell=False, capture=False)
2022-07-21 08:49:24,313 - util.py[DEBUG]: apt-update [eatmydata apt-get --option=Dpkg::Options::=--force-confold --option=Dpkg::options::=--force-unsafe-io --assume-yes --quiet update] took 15.602 seconds
2022-07-21 08:49:24,316 - helpers.py[DEBUG]: update-sources already ran (freq=once-per-instance)
2022-07-21 08:49:24,316 - debian.py[DEBUG]: Waiting for apt lock
2022-07-21 08:49:24,316 - debian.py[DEBUG]: apt lock available
2022-07-21 08:49:24,316 - subp.py[DEBUG]: Running command ['eatmydata', 'apt-get', '--option=Dpkg::Options::=--force-confold', '--option=Dpkg::options::=--force-unsafe-io', '--assume-yes', '--quiet', 'install', 'wireguard-tools'] with allowed return codes [0] (shell=False, capture=False)
2022-07-21 08:49:46,794 - util.py[DEBUG]: apt-install [eatmydata apt-get --option=Dpkg::Options::=--force-confold --option=Dpkg::options::=--force-unsafe-io --assume-yes --quiet install wireguard-tools] took 22.478 seconds
2022-07-21 08:49:46,795 - subp.py[DEBUG]: Running command lsmod with allowed return codes [0] (shell=True, capture=True)
2022-07-21 08:49:46,799 - cc_wireguard.py[DEBUG]: Loading wireguard kernel module
2022-07-21 08:49:46,799 - subp.py[DEBUG]: Running command modprobe wireguard with allowed return codes [0] (shell=True, capture=True)
2022-07-21 08:49:46,801 - util.py[WARNING]: Could not load wireguard module:
Unexpected error while running command.
Command: modprobe wireguard
Exit code: 1
Reason: -
Stdout:
Stderr: modprobe: FATAL: Module wireguard not found in directory /lib/modules/5.15.0-41-generic
2022-07-21 08:49:46,802 - util.py[DEBUG]: Could not load wireguard module:
Unexpected error while running command.
Command: modprobe wireguard
Exit code: 1
Reason: -
Stdout:
Stderr: modprobe: FATAL: Module wireguard not found in directory /lib/modules/5.15.0-41-generic
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/cloudinit/config/cc_wireguard.py", line 248, in load_wireguard_kernel_module
    subp.subp("modprobe wireguard", capture=True, shell=True)
  File "/usr/lib/python3/dist-packages/cloudinit/subp.py", line 336, in subp
    stdout=out, stderr=err, exit_code=rc, cmd=args
cloudinit.subp.ProcessExecutionError: Unexpected error while running command.
Command: modprobe wireguard
Exit code: 1
Reason: -
Stdout:
Stderr: modprobe: FATAL: Module wireguard not found in directory /lib/modules/5.15.0-41-generic
2022-07-21 08:49:46,802 - handlers.py[DEBUG]: finish: modules-config/config-wireguard: FAIL: running config-wireguard with frequency once-per-instance
...
```

aciba90 commented on cloudinit/config/cc_wireguard.py:
```diff
@@ -0,0 +1,290 @@
+    "title": "Module to configure Wireguard tunnel",
+    "description": MODULE_DESCRIPTION,
+    "distros": ["ubuntu"],
+    "frequency": PER_INSTANCE,
```
We have introduced a new schema property, in https://github.com/canonical/cloud-init/pull/1597, to skip the execution of a module if some config keys are not present. And I think this module can be skipped if `wireguard` is not given in the cloud config, right? If so, could you please rebase main and apply:

```suggestion
    "frequency": PER_INSTANCE,
    "activate_by_schema_keys": ["wireguard"],
```

aciba90 commented on tests/integration_tests/modules/test_wireguard.py:
```diff
@@ -0,0 +1,90 @@
+"""Integration test for the wireguard module.
+
+TODO Description
```
```suggestion
```

chifac08 commented:
> -e integration-tests tests/integration_tests/modules/test_wireguard.py

@aciba90 thx for reviewing. I implemented the changes you requested and figured out why loading wireguard module is failing. You can't load kernel modules in containers as they share the same kernel from the host system. It worked on my machine as my host loaded wireguard module. I decided to allow "lxd_vm" tests only as this makes more sense in this case. Unfortunately, with current integration tests using "lxd_container" as platform so it gets skipped. Another option would be to only allow "lxd_container" tests with loading the kernel module on the host (passing the lxd arguments via pytest).

Let me know what you think.

thx

holmanb commented:
> I decided to allow "lxd_vm" tests only as this makes more sense in this case. Unfortunately, with current integration tests using "lxd_container" as platform so it gets skipped. Another option would be to only allow "lxd_container" tests with loading the kernel module on the host (passing the lxd arguments via pytest).

Could we add coverage for the cloud marks to the integration test as well please? I think those should allow kmod insertion and are used in our pre-release testing (not github or travis CI).

Additionally, the package install test should still work in a container.

Maybe:
```suggestion
diff --git a/tests/integration_tests/modules/test_wireguard.py b/tests/integration_tests/modules/test_wireguard.py
index b5751bcf..27eae9f9 100644
--- a/tests/integration_tests/modules/test_wireguard.py
+++ b/tests/integration_tests/modules/test_wireguard.py
@@ -40,6 +40,11 @@ wireguard:
 @pytest.mark.ci
 @pytest.mark.user_data(USER_DATA)
 @pytest.mark.lxd_vm
+@pytest.mark.gce
+@pytest.mark.ec2
+@pytest.mark.azure
+@pytest.mark.openstack
+@pytest.mark.oci
 class TestWireguard:
     @pytest.mark.parametrize(
         "cmd,expected_out",
@@ -86,3 +91,12 @@ class TestWireguard:
     ):
         """Test that 'wg version' succeeds, indicating installation."""
         assert class_client.execute("wg version").ok
+
+
+@pytest.mark.lxd_container
+class TestWireguardWithoutKmod:
+    def test_wireguard_tools_installed(
+        self, class_client: IntegrationInstance
+    ):
+        """Test that 'wg version' succeeds, indicating installation."""
+        assert class_client.execute("wg version").ok
```

chifac08 commented:
> > I decided to allow "lxd_vm" tests only as this makes more sense in this case. Unfortunately, with current integration tests using "lxd_container" as platform so it gets skipped. Another option would be to only allow "lxd_container" tests with loading the kernel module on the host (passing the lxd arguments via pytest).
> 
> Could we add coverage for the cloud marks to the integration test as well please? I think those should allow kmod insertion and are used in our pre-release testing (not github or travis CI).
> 
> Additionally, the package install test should still work in a container.
> 
> Maybe something like this? (feel free to use this patch directly if want)
> 
> ```
> diff --git a/tests/integration_tests/modules/test_wireguard.py b/tests/integration_tests/modules/test_wireguard.py
> index b5751bcf..27eae9f9 100644
> --- a/tests/integration_tests/modules/test_wireguard.py
> +++ b/tests/integration_tests/modules/test_wireguard.py
> @@ -40,6 +40,11 @@ wireguard:
>  @pytest.mark.ci
>  @pytest.mark.user_data(USER_DATA)
>  @pytest.mark.lxd_vm
> +@pytest.mark.gce
> +@pytest.mark.ec2
> +@pytest.mark.azure
> +@pytest.mark.openstack
> +@pytest.mark.oci
>  class TestWireguard:
>      @pytest.mark.parametrize(
>          "cmd,expected_out",
> @@ -86,3 +91,12 @@ class TestWireguard:
>      ):
>          """Test that 'wg version' succeeds, indicating installation."""
>          assert class_client.execute("wg version").ok
> +
> +
> +@pytest.mark.lxd_container
> +class TestWireguardWithoutKmod:
> +    def test_wireguard_tools_installed(
> +        self, class_client: IntegrationInstance
> +    ):
> +        """Test that 'wg version' succeeds, indicating installation."""
> +        assert class_client.execute("wg version").ok
> ```

@holmanb i implemented your suggestions but unfortunately all tests get skipped and i don't why. As this module was only tested on ubuntu i also added
`@pytest.mark.ubuntu` to wireguard integrations tests. I hope you agree with me. Furthermore, i adjusted TRAVIS CI yaml by installing wireguard package as the test host must provide wireguard kernel module for "lxd_container" tests.

Please have a look at it. thx

holmanb commented:
Sorry for the delayed response @chifac08. I'm circling back to this review at the moment, currently testing.

One small request while I'm doing that: `doc/rtd/topics/modules.rst` will require an import of the module to pull in the wireguard docs.

holmanb commented on cloudinit/config/cc_wireguard.py:
```diff
@@ -0,0 +1,291 @@
+    installation.
+    """
+
+    packages = ["wireguard-tools"]
```
On Ubuntu I think we may need to install `resolvconf` for normal mode of operation on cloud-init, since `wg-quick` calls resolvconf, but it isn't guaranteed to be installed, since resolvconf isn't listed as a package dependency of wireguard (only suggested).

holmanb replied:
I suppose someone might want to use wireguard without dns settings, therefore not needing resolvconf. Feel free to ignore this suggestion.

@chifac08 Do you think it is worth mentioning that resolvconf would be required for dns in the docs? Let me know your thoughts.

chifac08 replied:
@holmanb i think we should document it but not install `resolvconf` on default. As most modern Linux distributions use systemd for DNS - `resolvectl` the user should be responsible for installing `resolvconf` package or just symlink it to `resolvectl`. I am going to add this to documentation.

holmanb replied:
@chifac08 Thanks, that makes sense to me!

holmanb commented on cloudinit/config/cc_wireguard.py:
```diff
@@ -0,0 +1,291 @@
+    @raises: RuntimeError for issues enabling WG interface.
+    """
+    try:
+        LOG.debug("Enabling wg-quick@%s at boot", {wg_int["name"]})
```
We have some funny-looking formatting throughout, specifically looking at the curly braces.
```
2022-08-03 23:16:42,598 - modules.py[DEBUG]: Running module wireguard (<module 'cloudinit.config.cc_wireguard' from '/usr/lib/python3/dist-packages/cloudinit/config/cc_wireguard.py'>) with frequency once-per-instance
2022-08-03 23:16:42,598 - handlers.py[DEBUG]: start: modules-config/config-wireguard: running config-wireguard with frequency once-per-instance
2022-08-03 23:16:42,598 - util.py[DEBUG]: Writing to /var/lib/cloud/instances/6d9d91a3-a659-41c5-8250-4ced8199065d/sem/config_wireguard - wb: [644] 24 bytes
2022-08-03 23:16:42,598 - helpers.py[DEBUG]: Running config-wireguard using lock (<FileLock using file '/var/lib/cloud/instances/6d9d91a3-a659-41c5-8250-4ced8199065d/sem/config_wireguard'>)
2022-08-03 23:16:42,598 - cc_wireguard.py[DEBUG]: Found Wireguard section in config
2022-08-03 23:16:42,598 - subp.py[DEBUG]: Running command lsmod with allowed return codes [0] (shell=True, capture=True)
2022-08-03 23:16:42,601 - cc_wireguard.py[DEBUG]: Configuring Wireguard interface {'wg0'}
2022-08-03 23:16:42,601 - cc_wireguard.py[DEBUG]: Writing wireguard config to file {'/etc/wireguard/wg0.conf'}
2022-08-03 23:16:42,601 - util.py[DEBUG]: Writing to /etc/wireguard/wg0.conf - wb: [600] 230 bytes
2022-08-03 23:16:42,601 - cc_wireguard.py[DEBUG]: Enabling wg-quick@{'wg0'} at boot
2022-08-03 23:16:42,601 - subp.py[DEBUG]: Running command ['systemctl', 'enable', 'wg-quick@wg0'] with allowed return codes [0] (shell=False, capture=True)
2022-08-03 23:16:42,917 - cc_wireguard.py[DEBUG]: Bringing up interface wg-quick@{'wg0'}
2022-08-03 23:16:42,918 - subp.py[DEBUG]: Running command ['systemctl', 'start', 'wg-quick@wg0'] with allowed return codes [0] (shell=False, capture=True)
2022-08-03 23:16:42,928 - cc_wireguard.py[DEBUG]: Running readinessprobe: '{'curl ifconfig.io > /etc/external.ip'}'
2022-08-03 23:16:42,928 - subp.py[DEBUG]: Running command curl ifconfig.io > /etc/external.ip with allowed return codes [0] (shell=True, capture=True)
2022-08-03 23:16:43,218 - handlers.py[DEBUG]: finish: modules-config/config-wireguard: SUCCESS: config-wireguard ran successfully
```
I'm guessing f-strings got left behind when switching to logging?

Could we fix those throughout the PR please?

holmanb reviewed (changes requested):
@chifac08 I just wrapped up some testing, I will leave some notes in this review of what I did for other reviewers, but this is looking pretty good to me. Once the comments I just added are addressed I'll test again and then give the +1. Thanks for all of your effort on this!


Manually tested:

Steps I took while testing:

1) Set up an EC2 wireguard server configured with forwarding enabled
2) Verify wireguard server is working with dev box, then disconnect dev box
3) In LXD VM install cloud-init deb built from this branch
4) Manually install resolvconf package
5) `cloud-init clean --logs --reboot`
6) verify connection during readiness probe and after 

config used:
```yaml
wireguard:
  interfaces:
    - name: wg0
      config_path: /etc/wireguard/wg0.conf
      content: |
         [Interface]
         Address = 10.0.0.2/24
         PrivateKey = <privkey>
         DNS = 1.1.1.1
         
         [Peer]
         PublicKey = <pubkey>
         Endpoint = <public ip>:51820
         AllowedIPs = 0.0.0.0/0, ::/0
  readinessprobe:
    - 'curl ifconfig.io > /etc/external.ip'
```

after booting the instance with cloud-init I see:
```
# cat /etc/external.ip 
<wireguard server public ip>
# wg
...
  transfer: 32.14 KiB received, 13.32 KiB sent
```

server config:
```
root@ip-172-31-42-92:/home/ubuntu# cat /etc/wireguard/wg0.conf 
[Interface]
Address = 10.0.0.1/24
SaveConfig = true
PostUp = iptables -A FORWARD -i wg0 -j ACCEPT; iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
PostDown = iptables -D FORWARD -i wg0 -j ACCEPT; iptables -t nat -D POSTROUTING -o eth0 -j MASQUERADE
ListenPort = 51820
PrivateKey = <privkey>

[Peer]
PublicKey = <pubkey>
AllowedIPs = 10.0.0.2/32
Endpoint = <vm public IP>:33604
```

logs:
```
2022-08-03 23:16:42,598 - modules.py[DEBUG]: Running module wireguard (<module 'cloudinit.config.cc_wireguard' from '/usr/lib/python3/dist-packages/cloudinit/config/cc_wireguard.py'>) with frequency once-per-instance
2022-08-03 23:16:42,598 - handlers.py[DEBUG]: start: modules-config/config-wireguard: running config-wireguard with frequency once-per-instance
2022-08-03 23:16:42,598 - util.py[DEBUG]: Writing to /var/lib/cloud/instances/6d9d91a3-a659-41c5-8250-4ced8199065d/sem/config_wireguard - wb: [644] 24 bytes
2022-08-03 23:16:42,598 - helpers.py[DEBUG]: Running config-wireguard using lock (<FileLock using file '/var/lib/cloud/instances/6d9d91a3-a659-41c5-8250-4ced8199065d/sem/config_wireguard'>)
2022-08-03 23:16:42,598 - cc_wireguard.py[DEBUG]: Found Wireguard section in config
2022-08-03 23:16:42,598 - subp.py[DEBUG]: Running command lsmod with allowed return codes [0] (shell=True, capture=True)
2022-08-03 23:16:42,601 - cc_wireguard.py[DEBUG]: Configuring Wireguard interface {'wg0'}
2022-08-03 23:16:42,601 - cc_wireguard.py[DEBUG]: Writing wireguard config to file {'/etc/wireguard/wg0.conf'}
2022-08-03 23:16:42,601 - util.py[DEBUG]: Writing to /etc/wireguard/wg0.conf - wb: [600] 230 bytes
2022-08-03 23:16:42,601 - cc_wireguard.py[DEBUG]: Enabling wg-quick@{'wg0'} at boot
2022-08-03 23:16:42,601 - subp.py[DEBUG]: Running command ['systemctl', 'enable', 'wg-quick@wg0'] with allowed return codes [0] (shell=False, capture=True)
2022-08-03 23:16:42,917 - cc_wireguard.py[DEBUG]: Bringing up interface wg-quick@{'wg0'}
2022-08-03 23:16:42,918 - subp.py[DEBUG]: Running command ['systemctl', 'start', 'wg-quick@wg0'] with allowed return codes [0] (shell=False, capture=True)
2022-08-03 23:16:42,928 - cc_wireguard.py[DEBUG]: Running readinessprobe: '{'curl ifconfig.io > /etc/external.ip'}'
2022-08-03 23:16:42,928 - subp.py[DEBUG]: Running command curl ifconfig.io > /etc/external.ip with allowed return codes [0] (shell=True, capture=True)
2022-08-03 23:16:43,218 - handlers.py[DEBUG]: finish: modules-config/config-wireguard: SUCCESS: config-wireguard ran successfully
```

chifac08 commented:
> @chifac08 I just wrapped up some testing, I will leave some notes in this review of what I did for other reviewers, but this is looking pretty good to me. Once the comments I just added are addressed I'll test again and then give the +1. Thanks for all of your effort on this!
> 
> Steps I took while testing:
> 
> 1. Set up an EC2 wireguard server configured with forwarding enabled
> 2. Verify wireguard server is working with dev box, then disconnect dev box
> 3. In LXD VM install cloud-init deb built from this branch
> 4. Manually install resolvconf package
> 5. `cloud-init clean --logs --reboot`
> 6. verify connection during readiness probe and after
> 
> config used:
> 
> ```yaml
> wireguard:
>   interfaces:
>     - name: wg0
>       config_path: /etc/wireguard/wg0.conf
>       content: |
>          [Interface]
>          Address = 10.0.0.2/24
>          PrivateKey = <privkey>
>          DNS = 1.1.1.1
>          
>          [Peer]
>          PublicKey = <pubkey>
>          Endpoint = <public ip>:51820
>          AllowedIPs = 0.0.0.0/0, ::/0
>   readinessprobe:
>     - 'curl ifconfig.io > /etc/external.ip'
> ```
> 
> after booting the instance with cloud-init I see:
> 
> ```
> # cat /etc/external.ip 
> <wireguard server public ip>
> # wg
> ...
>   transfer: 32.14 KiB received, 13.32 KiB sent
> ```
> 
> server config:
> 
> ```
> root@ip-172-31-42-92:/home/ubuntu# cat /etc/wireguard/wg0.conf 
> [Interface]
> Address = 10.0.0.1/24
> SaveConfig = true
> PostUp = iptables -A FORWARD -i wg0 -j ACCEPT; iptables -t nat -A POSTROUTING -o eth0 -j MASQUERADE
> PostDown = iptables -D FORWARD -i wg0 -j ACCEPT; iptables -t nat -D POSTROUTING -o eth0 -j MASQUERADE
> ListenPort = 51820
> PrivateKey = <privkey>
> 
> [Peer]
> PublicKey = <pubkey>
> AllowedIPs = 10.0.0.2/32
> Endpoint = <vm public IP>:33604
> ```
> 
> logs:
> 
> ```
> 2022-08-03 23:16:42,598 - modules.py[DEBUG]: Running module wireguard (<module 'cloudinit.config.cc_wireguard' from '/usr/lib/python3/dist-packages/cloudinit/config/cc_wireguard.py'>) with frequency once-per-instance
> 2022-08-03 23:16:42,598 - handlers.py[DEBUG]: start: modules-config/config-wireguard: running config-wireguard with frequency once-per-instance
> 2022-08-03 23:16:42,598 - util.py[DEBUG]: Writing to /var/lib/cloud/instances/6d9d91a3-a659-41c5-8250-4ced8199065d/sem/config_wireguard - wb: [644] 24 bytes
> 2022-08-03 23:16:42,598 - helpers.py[DEBUG]: Running config-wireguard using lock (<FileLock using file '/var/lib/cloud/instances/6d9d91a3-a659-41c5-8250-4ced8199065d/sem/config_wireguard'>)
> 2022-08-03 23:16:42,598 - cc_wireguard.py[DEBUG]: Found Wireguard section in config
> 2022-08-03 23:16:42,598 - subp.py[DEBUG]: Running command lsmod with allowed return codes [0] (shell=True, capture=True)
> 2022-08-03 23:16:42,601 - cc_wireguard.py[DEBUG]: Configuring Wireguard interface {'wg0'}
> 2022-08-03 23:16:42,601 - cc_wireguard.py[DEBUG]: Writing wireguard config to file {'/etc/wireguard/wg0.conf'}
> 2022-08-03 23:16:42,601 - util.py[DEBUG]: Writing to /etc/wireguard/wg0.conf - wb: [600] 230 bytes
> 2022-08-03 23:16:42,601 - cc_wireguard.py[DEBUG]: Enabling wg-quick@{'wg0'} at boot
> 2022-08-03 23:16:42,601 - subp.py[DEBUG]: Running command ['systemctl', 'enable', 'wg-quick@wg0'] with allowed return codes [0] (shell=False, capture=True)
> 2022-08-03 23:16:42,917 - cc_wireguard.py[DEBUG]: Bringing up interface wg-quick@{'wg0'}
> 2022-08-03 23:16:42,918 - subp.py[DEBUG]: Running command ['systemctl', 'start', 'wg-quick@wg0'] with allowed return codes [0] (shell=False, capture=True)
> 2022-08-03 23:16:42,928 - cc_wireguard.py[DEBUG]: Running readinessprobe: '{'curl ifconfig.io > /etc/external.ip'}'
> 2022-08-03 23:16:42,928 - subp.py[DEBUG]: Running command curl ifconfig.io > /etc/external.ip with allowed return codes [0] (shell=True, capture=True)
> 2022-08-03 23:16:43,218 - handlers.py[DEBUG]: finish: modules-config/config-wireguard: SUCCESS: config-wireguard ran successfully
> ```

@holmanb implemented changes you requested. Additionally, i reformatted documentation layout for wireguard module. Thank you for reviewing and the open-minded atmosphere. It was a pleasure developing this module with you guys and I hope that it will make it's way in the upstream. 

If there are any other suggestions/changes, please let me know.

holmanb commented:
Retested, logs:

```python
2022-08-05 22:09:58,235 - modules.py[DEBUG]: Running module wireguard (<module 'cloudinit.config.cc_wireguard' from '/usr/lib/python3/dist-packages/cloudinit/config/cc_wireguard.py'>) with frequency once-per-instance
2022-08-05 22:09:58,235 - handlers.py[DEBUG]: start: modules-config/config-wireguard: running config-wireguard with frequency once-per-instance
2022-08-05 22:09:58,235 - util.py[DEBUG]: Writing to /var/lib/cloud/instances/6d9d91a3-a659-41c5-8250-4ced8199065d/sem/config_wireguard - wb: [644] 24 bytes
2022-08-05 22:09:58,235 - helpers.py[DEBUG]: Running config-wireguard using lock (<FileLock using file '/var/lib/cloud/instances/6d9d91a3-a659-41c5-8250-4ced8199065d/sem/config_wireguard'>)
2022-08-05 22:09:58,235 - cc_wireguard.py[DEBUG]: Found Wireguard section in config
2022-08-05 22:09:58,235 - subp.py[DEBUG]: Running command lsmod with allowed return codes [0] (shell=True, capture=True)
2022-08-05 22:09:58,248 - cc_wireguard.py[DEBUG]: Configuring Wireguard interface wg0
2022-08-05 22:09:58,248 - cc_wireguard.py[DEBUG]: Writing wireguard config to file /etc/wireguard/wg0.conf
2022-08-05 22:09:58,248 - util.py[DEBUG]: Writing to /etc/wireguard/wg0.conf - wb: [600] 230 bytes
2022-08-05 22:09:58,248 - cc_wireguard.py[DEBUG]: Enabling wg-quick@wg0 at boot
2022-08-05 22:09:58,249 - subp.py[DEBUG]: Running command ['systemctl', 'enable', 'wg-quick@wg0'] with allowed return codes [0] (shell=False, capture=True)
2022-08-05 22:09:58,561 - cc_wireguard.py[DEBUG]: Bringing up interface wg-quick@wg0
2022-08-05 22:09:58,562 - subp.py[DEBUG]: Running command ['systemctl', 'start', 'wg-quick@wg0'] with allowed return codes [0] (shell=False, capture=True)
2022-08-05 22:09:58,596 - cc_wireguard.py[DEBUG]: Running readinessprobe: 'curl ifconfig.io > /etc/external.ip'
2022-08-05 22:09:58,596 - subp.py[DEBUG]: Running command curl ifconfig.io > /etc/external.ip with allowed return codes [0] (shell=True, capture=True)
2022-08-05 22:09:58,897 - handlers.py[DEBUG]: finish: modules-config/config-wireguard: SUCCESS: config-wireguard ran successfully
```

holmanb reviewed (approved):
> @holmanb implemented changes you requested. Additionally, i reformatted documentation layout for wireguard module.

Docs look good to me. Nice additions and changes.


> Thank you for reviewing and the open-minded atmosphere. It was a pleasure developing this module with you guys and I hope that it will make it's way in the upstream.

Likewise :) The effort is much appreciated. Great job!


> If there are any other suggestions/changes, please let me know.

No more requests from me.

chifac08 commented:
@aciba90 As cloud-init documentation states that "no “Request changes” reviews from any Reviewer" is mandatory to a land a PR, I would be delighted if you could review my implementation based on your requested changes/suggestions. 
Thank you

aciba90 reviewed (approved):
> @aciba90 As cloud-init documentation states that "no “Request changes” reviews from any Reviewer" is mandatory to a land a PR, I would be delighted if you could review my implementation based on your requested changes/suggestions.
Thank you

LGTM. Thanks, @chifac08, for contributing this new module!

blackboxsw reviewed (dismissed):
Thanks for this good work here.
Related to this new module, we'll need to also add a corresponding expectation that the module itself is disabled if no related #cloud-config directives "wireguard" are present.

```diff
diff --git a/tests/integration_tests/modules/test_ca_certs.py b/tests/integration_tests/modules/test_ca_certs.py
index 9e095e4d4..8d18fb765 100644
--- a/tests/integration_tests/modules/test_ca_certs.py
+++ b/tests/integration_tests/modules/test_ca_certs.py
@@ -125,6 +125,7 @@ class TestCaCerts:
             "ubuntu-advantage",
             "ubuntu-drivers",
             "update_etc_hosts",
+            "wireguard",
             "write-files",
             "write-files-deferred",
         }
```

blackboxsw reviewed (commented):
Thank your @chifac08 I dismiss my needs fixing

chifac08 commented:
@blackboxsw implemented it :) Anything else i should take care of?
thx

holmanb commented:
@chifac08 There are no outstanding issues and we have enough approvals to merge, so I will go ahead and do that (blackboxsw is unavailable for further review at this time).

Since this introduces a new config module I will take the liberty of including the additional context in commit history (and rework the commit message a bit).

Thanks again :)
